Mathlib/LinearAlgebra/LinearProgramming.lean

```diff
@@ -0,0 +1,104 @@
+/-
+Copyright (c) 2023 Martin Dvorak. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Dvorak
+-/
+import Mathlib.LinearAlgebra.Matrix.ToLin
+import Mathlib.LinearAlgebra.AffineSpace.AffineMap
+
+/-! # Linear programming
+
+Minimizing a linear function on a region defined by linear inequalities.
+
+## Main definitions
+
+ * `LinearProgram` defines a linear program in a form that generalizes "A x ≥ b".
+ * `LinearProgram.feasibles` is the set of all admissible solutions to given linear program.
+ * `LinearProgram.MinAt` defines an optimum solution of given linear program.
+
+-/
+
+/-- Linear program in the form of inequalities with general variables. -/
+structure LinearProgram (K : Type*) {V : Type*} (P : Type*)
+    [LinearOrderedField K] [AddCommGroup V] [Module K V] [AddTorsor V P] where
+  /-- Inequality constraints (given in the form "aᵀx - b ≥ 0") -/
+  constraints : List (P →ᵃ[K] K)
+  /-- The objective function (affine map) -/
+  objective : P →ᵃ[K] K
+
+variable {K V P : Type*} [LinearOrderedField K] [AddCommGroup V] [Module K V] [AddTorsor V P]
+
+/-- Constructs a linear program given a list of equalities, a list of inequalities,
+    and an objective function. -/
+def LinearProgram.mkOfEqs
+    (equalities inequalities : List (P →ᵃ[K] K)) (objective : P →ᵃ[K] K) :
+    LinearProgram K P :=
+  { constraints := inequalities ++ equalities ++ equalities.map Neg.neg, objective }
+
+/-- The set of all admissible solutions to given linear program. -/
+def LinearProgram.feasibles (lp : LinearProgram K P) : Set P :=
+  {x | ∀ ⦃a⦄, a ∈ lp.constraints → 0 ≤ a x}
+
+@[simp] lemma LinearProgram.mem_feasibles {lp : LinearProgram K P} {x : P} :
+    x ∈ lp.feasibles ↔ ∀ ⦃a⦄, a ∈ lp.constraints → 0 ≤ a x :=
+  Iff.rfl
+
+/-- Given linear program is minimized at given point. -/
+def LinearProgram.MinAt (lp : LinearProgram K P) (x : P) : Prop :=
+  IsLeast (lp.objective '' lp.feasibles) (lp.objective x)
+
+lemma LinearProgram.feasibles_mkOfEqs
+    (equalities inequalities : List (P →ᵃ[K] K)) (objective : P →ᵃ[K] K) :
+    (mkOfEqs equalities inequalities objective).feasibles =
+    { x : P | (∀ a ∈ equalities, a x = 0) ∧ (∀ a ∈ inequalities, 0 ≤ a x) } := by
+  ext x
+  constructor
+  · intro hyp
+    rw [Set.mem_setOf_eq]
+    simp only [LinearProgram.feasibles, LinearProgram.mkOfEqs,
+      List.append_assoc, List.mem_append, List.mem_map, Set.mem_setOf_eq,
+      Function.Involutive.exists_mem_and_apply_eq_iff, neg_involutive] at hyp
+    constructor
+    · intro constr_eq mem_equalities
+      refine le_antisymm ?neg (hyp (by simp [mem_equalities]))
+      rw [← Left.nonneg_neg_iff, ← Pi.neg_apply, ← AffineMap.coe_neg]
+      apply hyp
+      simp [mem_equalities]
+    · intro constr_le mem_inequalities
+      exact hyp (by simp [mem_inequalities])
+  · intro hyp
+    simp only [Set.mem_setOf_eq, LinearProgram.feasibles] at hyp
```

Comment:
```suggestion
    simp only [mem_feasibles] at hyp
```
You shouldn't need to ever unfold `feasibles` any more. I recommend also doing this at the start of the proof before you split the goals

Reply:
Unresolving for my second sentence

Reply:
Yeah, I haven't opened in IDE yet.

Reply:
Now.